src/c++/tests/memory_leak_test.cc

```diff
@@ -105,70 +105,93 @@ ValidateResult(
   std::cout << result->DebugString() << std::endl;
 }
 
+void
+ValidateResponse(
+    std::shared_ptr<tc::InferResult> results_ptr,
+    const std::vector<int32_t>& input0_data)
+{
+  // Validate results
+  if (results_ptr->RequestStatus().IsOk()) {
+    ValidateResult(results_ptr, input0_data);
+  } else {
+    std::cerr << "error: Inference failed: " << results_ptr->RequestStatus()
+              << std::endl;
+    exit(1);
+  }
+}
 
+template <typename Client>
 void
-RunSynchronousInference(
+InferWithRetries(
+    const std::unique_ptr<Client>& client, tc::InferResult** results,
+    tc::InferOptions& options, std::vector<tc::InferInput*>& inputs,
+    std::vector<const tc::InferRequestedOutput*>& outputs)
+{
+  // Exit early if we succeed first try
+  auto err = client->Infer(results, options, inputs, outputs);
+  if (err.IsOk()) {
+    return;
+  }
+
+  // If the host runs out of available sockets due to TIME_WAIT, sleep and
+  // retry on failure to give time for sockets to become available.
+  int max_retries = 5;
+  int sleep_secs = 60;
+  bool success = false;
+  for (int i = 0; i < max_retries; i++) {
+    std::cerr << "Error: " << err << std::endl;
+    std::cerr << "Sleeping for " << sleep_secs
+              << " seconds and retrying. [Attempt: " << i + 1 << "/"
+              << max_retries << "]" << std::endl;
+    sleep(sleep_secs);
+
+    err = client->Infer(results, options, inputs, outputs);
+    if (err.IsOk()) {
+      success = true;
+      break;
+    }
+  }
+
+  if (!success) {
+    std::cerr << "error: Exceeded max tries [" << max_retries
+              << "] on inference without success" << std::endl;
+    exit(1);
+  }
+}
+
+// Client should be tc::InferenceServerHttpClient or
+// tc::InferenceServerGrpcClient
+template <typename Client>
+void
+RunSyncInfer(
     std::vector<tc::InferInput*>& inputs,
     std::vector<const tc::InferRequestedOutput*>& outputs,
     tc::InferOptions& options, std::vector<int32_t>& input0_data, bool reuse,
-    std::string url, bool verbose, std::string protocol, uint32_t repetitions)
+    std::string url, bool verbose, uint32_t repetitions)
 {
   // If re-use is enabled then use these client objects else use new objects for
   // each inference request.
-  std::unique_ptr<tc::InferenceServerGrpcClient> grpc_client_reuse;
-  std::unique_ptr<tc::InferenceServerHttpClient> http_client_reuse;
+  std::unique_ptr<Client> client_reuse;
```

Comment:
Could simplify the reuse logic by just keeping one client and resetting if reuse is not enabled. 

Open Question: I'm tempted to say we should follow the same logic independent of reuse or no reuse - i.e. run with retries.  That maybe requires a discussion on how to deal with potentially hiding errors. As this test is specifically about memory growth - I'm less concerned about retries. If consensus is that retries then maybe we slow down the test - I think even 1 second sleep should do it at 20,000 requests.

```
 // If re-use is enabled then use these client objects else use new objects for
 165 │   // each inference request.
 166 │   std::unique_ptr<ClientType> client;
 167 │   FAIL_IF_ERR(Client::Create(&client, url, verbose), "unable to create client");
 168 │
 169 │   for (size_t i = 0; i < repetitions; ++i) {
 170 │     tc::InferResult* results;
 171 │
 172 │     InferWithRetries<ClientType>(client,&results,inputs,outputs);
 173 │
 174 │     if (!reuse) {
 175 │       FAIL_IF_ERR(ClientType::Create(&client, url, verbose), "unable to create client");
 176 │       InferWithRetries<ClientType>(client, &results, options, inputs, outputs);
 177 │     }
 178 │
 179 │     std::shared_ptr<tc::InferResult> results_ptr(results);
 180 │     ValidateResponse(results_ptr, input0_data);
 181 │   }

```

Reply:
Good idea, done